File: rust-agent/cli/src/session-utils.ts
export interface Message {
  role: string;
  content: string;
}

export function transformMessages(
  apiMessages: Array<{ role: string; content: any }>,
): Message[] {
  const result: Message[] = [];

  for (const msg of apiMessages) {
    if (msg.role === "user") {
      if (typeof msg.content === "string") {
        result.push({ role: "user", content: msg.content });
      } else if (Array.isArray(msg.content)) {
        let texts: string[] = [];
        for (const block of msg.content) {
          if (block?.type === "tool_result") {
            if (texts.length > 0) {
              result.push({ role: "user", content: texts.join("") });
              texts = [];
            }
            result.push({
              role: "tool_result",
              content: String(block.content ?? ""),
            });
          } else if (block?.type === "text" && typeof block.text === "string") {
            texts.push(block.text);
          }
        }
        if (texts.length > 0) {
          result.push({ role: "user", content: texts.join("") });
        }
      }
    } else if (msg.role === "assistant") {
      if (typeof msg.content === "string") {
        result.push({ role: "assistant", content: msg.content });
      } else if (Array.isArray(msg.content)) {
        for (const block of msg.content) {
          if (block?.type === "text" && typeof block.text === "string") {
            result.push({ role: "assistant", content: block.text });
          } else if (block?.type === "tool_use") {
            result.push({
              role: "tool_call",
              content: JSON.stringify({
                name: block.name,
                input: block.input,
              }),
            });
          }
        }
      }
    } else {
      console.warn(`Unknown role in history: ${msg.role}`);
    }
  }

  return result;
}

File: rust-agent/cli_py/rust_agent_cli/session_utils.py
"""消息转换工具 — 复刻 cli/src/session-utils.ts."""

from dataclasses import dataclass
from typing import Any


@dataclass
class Message:
    role: str
    content: str


def transform_messages(api_messages: list[dict[str, Any]]) -> list[Message]:
    result: list[Message] = []
    for msg in api_messages:
        role = msg.get("role")
        content = msg.get("content")
        if role == "user":
            if isinstance(content, str):
                result.append(Message(role="user", content=content))
            elif isinstance(content, list):
                texts: list[str] = []
                for block in content:
                    if block.get("type") == "tool_result":
                        if texts:
                            result.append(Message(role="user", content="".join(texts)))
                            texts = []
                        result.append(
                            Message(role="tool_result", content=str(block.get("content", "")))
                        )
                    elif block.get("type") == "text" and isinstance(block.get("text"), str):
                        texts.append(block["text"])
                if texts:
                    result.append(Message(role="user", content="".join(texts)))
        elif role == "assistant":
            if isinstance(content, str):
                result.append(Message(role="assistant", content=content))
            elif isinstance(content, list):
                for block in content:
                    if block.get("type") == "text" and isinstance(block.get("text"), str):
                        result.append(Message(role="assistant", content=block["text"]))
                    elif block.get("type") == "tool_use":
                        result.append(
                            Message(
                                role="tool_call",
                                content=__import__("json").dumps(
                                    {"name": block.get("name"), "input": block.get("input")}
                                ),
                            )
                        )
    return result
